### PR TITLE
[inductor] track provenance of masks from indices

### LINKED ISSUE #100530: Inductor generates incorrect indexing code with an arange
### 🐛 Describe the bug

```
import torch._inductor.overrides

import torch
from torch import tensor, device
import torch.fx as fx
from torch._dynamo.testing import rand_strided
from math import inf
from torch.fx.experimental.proxy_tensor import make_fx

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.load_config(b'\x80\x02}q\x00(X\r\x00\x00\x00log_file_nameq\x01NX\x07\x00\x00\x00verboseq\x02\x89X\x12\x00\x00\x00verify_correctnessq\x03\x89X\x12\x00\x00\x00minimum_call_countq\x04K\x01X\x15\x00\x00\x00dead_code_eliminationq\x05\x88X\x10\x00\x00\x00cache_size_limitq\x06K@X\x0e\x00\x00\x00specialize_intq\x07\x89X\x0e\x00\x00\x00dynamic_shapesq\x08\x89X\x18\x00\x00\x00assume_static_by_defaultq\t\x88X\x18\x00\x00\x00automatic_dynamic_shapesq\n\x88X\x19\x00\x00\x00allow_ignore_mark_dynamicq\x0b\x89X\x10\x00\x00\x00guard_nn_modulesq\x0c\x89X\x1b\x00\x00\x00traceable_tensor_subclassesq\rc__builtin__\nset\nq\x0e]q\x0f\x85q\x10Rq\x11X\x0f\x00\x00\x00suppress_errorsq\x12\x89X\x15\x00\x00\x00replay_record_enabledq\x13\x89X \x00\x00\x00rewrite_assert_with_torch_assertq\x14\x88X\x12\x00\x00\x00print_graph_breaksq\x15\x89X\x15\x00\x00\x00print_specializationsq\x16\x89X\x19\x00\x00\x00summarize_dim_constraintsq\x17\x89X\x07\x00\x00\x00disableq\x18\x89X*\x00\x00\x00allowed_functions_module_string_ignorelistq\x19h\x0e]q\x1a(X\r\x00\x00\x00torch._decompq\x1bX\r\x00\x00\x00torch.testingq\x1cX\x0c\x00\x00\x00torch._primsq\x1dX\x13\x00\x00\x00torch.distributionsq\x1eX\x0b\x00\x00\x00torch._refsq\x1fe\x85q Rq!X\x12\x00\x00\x00repro_forward_onlyq"\x89X\x0f\x00\x00\x00repro_toleranceq#G?PbM\xd2\xf1\xa9\xfcX\x16\x00\x00\x00capture_scalar_outputsq$\x89X \x00\x00\x00capture_dynamic_output_shape_opsq%\x89X\x19\x00\x00\x00enforce_cond_guards_matchq&\x88X\x0c\x00\x00\x00optimize_ddpq\'\x88X\x10\x00\x00\x00skip_fsdp_guardsq(\x88X\x19\x00\x00\x00skip_nnmodule_hook_guardsq)\x88X\x1a\x00\x00\x00raise_on_ctx_manager_usageq*\x88X\x1c\x00\x00\x00raise_on_unsafe_aot_autogradq+\x89X\x17\x00\x00\x00raise_on_backend_changeq,\x89X\x18\x00\x00\x00error_on_nested_fx_traceq-\x88X\t\x00\x00\x00allow_rnnq.\x89X\x12\x00\x00\x00error_on_recompileq/\x89X\x08\x00\x00\x00base_dirq0X\x1d\x00\x00\x00/scratch/chilli/fresh/pytorchq1X\x17\x00\x00\x00numpy_ndarray_as_tensorq2\x89X\x12\x00\x00\x00DEBUG_DIR_VAR_NAMEq3X\x17\x00\x00\x00TORCH_COMPILE_DEBUG_DIRq4X\x0e\x00\x00\x00debug_dir_rootq5X)\x00\x00\x00/scratch/chilli/fresh/torch_compile_debugq6X\x13\x00\x00\x00_save_config_ignoreq7h\x0e]q8(X!\x00\x00\x00skipfiles_inline_module_allowlistq9X\x0b\x00\x00\x00repro_levelq:X\x0b\x00\x00\x00repro_afterq;X\x12\x00\x00\x00constant_functionsq<e\x85q=Rq>u.')
torch._inductor.config.load_config(b'\x80\x02}q\x00(X\x05\x00\x00\x00debugq\x01\x89X\x10\x00\x00\x00disable_progressq\x02\x88X\x10\x00\x00\x00verbose_progressq\x03\x89X\x0b\x00\x00\x00cpp_wrapperq\x04\x89X\x03\x00\x00\x00dceq\x05\x89X\x14\x00\x00\x00static_weight_shapesq\x06\x88X\x0c\x00\x00\x00size_assertsq\x07\x88X\x10\x00\x00\x00pick_loop_ordersq\x08\x88X\x0f\x00\x00\x00inplace_buffersq\t\x88X\x11\x00\x00\x00benchmark_harnessq\n\x88X\x0f\x00\x00\x00epilogue_fusionq\x0b\x88X\x15\x00\x00\x00epilogue_fusion_firstq\x0c\x89X\x0f\x00\x00\x00pattern_matcherq\r\x88X\x13\x00\x00\x00split_cat_fx_passesq\x0e\x88X\n\x00\x00\x00reorderingq\x0f\x89X\x0c\x00\x00\x00max_autotuneq\x10\x89X\x16\x00\x00\x00max_autotune_pointwiseq\x11\x89X\x11\x00\x00\x00max_autotune_gemmq\x12\x89X\x15\x00\x00\x00search_autotune_cacheq\x13\x89X\x13\x00\x00\x00autotune_in_subprocq\x14\x89X\x19\x00\x00\x00coordinate_descent_tuningq\x15\x89X\x17\x00\x00\x00realize_reads_thresholdq\x16K\x04X\x17\x00\x00\x00realize_bytes_thresholdq\x17M\xd0\x07X\x1b\x00\x00\x00realize_acc_reads_thresholdq\x18K\x08X\x0f\x00\x00\x00fallback_randomq\x19\x89X\x12\x00\x00\x00implicit_fallbacksq\x1a\x88X\x11\x00\x00\x00aggressive_fusionq\x1b\x89X\x0f\x00\x00\x00max_fusion_sizeq\x1cK@X\x1b\x00\x00\x00unroll_reductions_thresholdq\x1dK\x08X\x0e\x00\x00\x00comment_originq\x1e\x89X\x0e\x00\x00\x00conv_1x1_as_mmq\x1f\x89X\x10\x00\x00\x00split_reductionsq \x88X\x0e\x00\x00\x00lowmem_dropoutq!\x88X\x10\x00\x00\x00benchmark_kernelq"\x89X\x14\x00\x00\x00is_nightly_or_sourceq#\x88X\x12\x00\x00\x00developer_warningsq$\x88X\x0f\x00\x00\x00compile_threadsq%K X\x10\x00\x00\x00global_cache_dirq&NX\x13\x00\x00\x00kernel_name_max_opsq\'K\nX\r\x00\x00\x00shape_paddingq(\x89X\x0e\x00\x00\x00permute_fusionq)\x89X\x1a\x00\x00\x00profiler_mark_wrapper_callq*\x89X\x18\x00\x00\x00_raise_error_for_testingq+\x89X\x0c\x00\x00\x00_profile_varq,X\x00\x00\x00\x00q-X\x11\x00\x00\x00profile_bandwidthq.\x89X\x17\x00\x00\x00profile_bandwidth_regexq/h-X\x13\x00\x00\x00disable_cpp_codegenq0\x89X\x0b\x00\x00\x00cpp.threadsq1J\xff\xff\xff\xffX\x16\x00\x00\x00cpp.no_redundant_loopsq2\x88X\x13\x00\x00\x00cpp.dynamic_threadsq3\x89X\x0b\x00\x00\x00cpp.simdlenq4NX\x12\x00\x00\x00cpp.min_chunk_sizeq5M\x00\x10X\x07\x00\x00\x00cpp.cxxq6NX\x03\x00\x00\x00g++q7\x86q8X\x19\x00\x00\x00cpp.enable_kernel_profileq9\x89X\x12\x00\x00\x00cpp.weight_prepackq:\x88X\x11\x00\x00\x00triton.cudagraphsq;\x89X\x16\x00\x00\x00triton.cudagraph_treesq<\x88X"\x00\x00\x00triton.slow_path_cudagraph_assertsq=\x89X"\x00\x00\x00triton.fast_path_cudagraph_assertsq>\x89X\x1c\x00\x00\x00triton.skip_cudagraph_warmupq?\x89X\x17\x00\x00\x00triton.debug_sync_graphq@\x89X\x18\x00\x00\x00triton.debug_sync_kernelqA\x89X\x15\x00\x00\x00triton.dense_indexingqB\x89X\x10\x00\x00\x00triton.max_tilesqCK\x02X\x19\x00\x00\x00triton.autotune_pointwiseqD\x88X\'\x00\x00\x00triton.tiling_prevents_pointwise_fusionqE\x88X\'\x00\x00\x00triton.tiling_prevents_reduction_fusionqF\x88X\x1f\x00\x00\x00triton.assert_indirect_indexingqG\x88X\x1a\x00\x00\x00triton.unique_kernel_namesqH\x89X\x18\x00\x00\x00triton.descriptive_namesqIX\r\x00\x00\x00original_atenqJX\x1c\x00\x00\x00triton.persistent_reductionsqK\x88X\x16\x00\x00\x00triton.divisible_by_16qL\x88X\x10\x00\x00\x00triton.max_blockqM}qN(X\x01\x00\x00\x00XqOM\x00\x08X\x01\x00\x00\x00YqPM\x00\x04X\x01\x00\x00\x00ZqQM\x00\x04uX\x12\x00\x00\x00triton.store_cubinqR\x89X\x16\x00\x00\x00triton.spill_thresholdqSK\x00X\r\x00\x00\x00trace.enabledqT\x89X\x0f\x00\x00\x00trace.debug_logqU\x89X\x0e\x00\x00\x00trace.info_logqV\x89X\x0e\x00\x00\x00trace.fx_graphqW\x88X\x1a\x00\x00\x00trace.fx_graph_transformedqX\x88X\x13\x00\x00\x00trace.ir_pre_fusionqY\x88X\x14\x00\x00\x00trace.ir_post_fusionqZ\x88X\x11\x00\x00\x00trace.output_codeq[\x88X\x13\x00\x00\x00trace.graph_diagramq\\\x89X\x15\x00\x00\x00trace.compile_profileq]\x89X\x13\x00\x00\x00_save_config_ignoreq^c__builtin__\nset\nq_]q`X\x10\x00\x00\x00trace.upload_tarqaa\x85qbRqcu.')
torch._functorch.config.load_config(b'\x80\x02}q\x00(X\x15\x00\x00\x00functionalize_rng_opsq\x01\x89X\x16\x00\x00\x00fake_tensor_allow_metaq\x02\x88X\x0c\x00\x00\x00debug_assertq\x03\x89X\x11\x00\x00\x00debug_partitionerq\x04\x89X\x14\x00\x00\x00static_weight_shapesq\x05\x88X\x03\x00\x00\x00cseq\x06\x88X\x10\x00\x00\x00max_dist_from_bwq\x07K\nu.')


# REPLACEABLE COMMENT FOR TESTING PURPOSES


# torch version: 2.1.0a0+gitd1fbd33
# torch cuda version: 11.8
# torch git version: d1fbd33c70205d7567c9e8f070bbbd5fc22d4a41


# CUDA Info:
# nvcc: NVIDIA (R) Cuda compiler driver
# Copyright (c) 2005-2022 NVIDIA Corporation
# Built on Wed_Sep_21_10:33:58_PDT_2022
# Cuda compilation tools, release 11.8, V11.8.89
# Build cuda_11.8.r11.8/compiler.31833905_0

# GPU Hardware Info:
# NVIDIA A100-SXM4-40GB : 2


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, arg450_1, arg451_1, div_16, clone_41):
        add_238 = torch.ops.aten.add.Tensor(clone_41, div_16);  clone_41 = div_16 = None
        convert_element_type_239 = torch.ops.prims.convert_element_type.default(add_238, torch.float32);  add_238 = None
        iota_3 = torch.ops.prims.iota.default(48, start = 0, step = 1, dtype = torch.int64, device = device(type='cuda', index=0), requires_grad = False)
        convert_element_type_240 = torch.ops.prims.convert_element_type.default(iota_3, torch.float32);  iota_3 = None
        mul_241 = torch.ops.aten.mul.Tensor(convert_element_type_240, 1);  convert_element_type_240 = None
        add_239 = torch.ops.aten.add.Tensor(mul_241, 0);  mul_241 = None
        mul_242 = torch.ops.aten.mul.Tensor(add_239, 0.5);  add_239 = None
        convert_element_type_241 = torch.ops.prims.convert_element_type.default(mul_242, torch.int64);  mul_242 = None
        unsqueeze_288 = torch.ops.aten.unsqueeze.default(convert_element_type_241, -1);  convert_element_type_241 = None
        iota_4 = torch.ops.prims.iota.default(48, start = 0, step = 1, dtype = torch.int64, device = device(type='cuda', index=0), requires_grad = False)
        convert_element_type_242 = torch.ops.prims.convert_element_type.default(iota_4, torch.float32);  iota_4 = None
        mul_243 = torch.ops.aten.mul.Tensor(convert_element_type_242, 1);  convert_element_type_242 = None
        add_240 = torch.ops.aten.add.Tensor(mul_243, 0);  mul_243 = None
        mul_244 = torch.ops.aten.mul.Tensor(add_240, 0.5);  add_240 = None
        convert_element_type_243 = torch.ops.prims.convert_element_type.default(mul_244, torch.int64);  mul_244 = None
        slice_41 = torch.ops.aten.slice.Tensor(convert_element_type_239, 0, 0, 9223372036854775807);  convert_element_type_239 = None
        slice_42 = torch.ops.aten.slice.Tensor(slice_41, 1, 0, 9223372036854775807);  slice_41 = None
        index_1 = torch.ops.aten.index.Tensor(slice_42, [None, None, unsqueeze_288, convert_element_type_243]);  slice_42 = unsqueeze_288 = convert_element_type_243 = None
        convert_element_type_244 = torch.ops.prims.convert_element_type.default(index_1, torch.float16);  index_1 = None
        convolution_45 = torch.ops.aten.convolution.default(convert_element_type_244, arg450_1, arg451_1, [1, 1], [1, 1], [1, 1], False, [0, 0], 1);  convert_element_type_244 = arg450_1 = arg451_1 = None
        return (convolution_45,)

import torch._dynamo.repro.after_aot
reader = torch._dynamo.repro.after_aot.InputReader(save_dir='/scratch/chilli/fresh/torch_compile_debug/run_2023_05_03_06_21_37_927237-pid_2928382/minifier/checkpoints')
buf0 = reader.storage('0ad6a8ab63587fb57b8a9a6ff5b4f285fd9a652d', 29491200, device=device(type='cuda', index=0), dtype_hint=torch.float16)
t0 = reader.tensor(buf0, (1280, 1280, 3, 3), (11520, 1, 3840, 1280), dtype=torch.float16)
buf1 = reader.storage('6925e8962559459dc60df50bfe4474b7251f3c89', 2560, device=device(type='cuda', index=0), dtype_hint=torch.float16)
t1 = reader.tensor(buf1, (1280,), dtype=torch.float16)
buf2 = reader.storage('2b4e027b52b6e5429227db749f0a002b0c5e51ae', 2949120, device=device(type='cuda', index=0), dtype_hint=torch.float16)
t2 = reader.tensor(buf2, (2, 1280, 24, 24), (737280, 1, 30720, 1280), dtype=torch.float16)
buf3 = reader.storage('2b4e027b52b6e5429227db749f0a002b0c5e51ae', 2949120, device=device(type='cuda', index=0), dtype_hint=torch.float16)
t3 = reader.tensor(buf3, (2, 1280, 24, 24), dtype=torch.float16)
args = [t0, t1, t2, t3]
mod = make_fx(Repro(), tracing_mode='real')(*args)

from torch._inductor.compile_fx import compile_fx_inner
from torch._dynamo.debug_utils import same_two_models

compiled = compile_fx_inner(mod, args)
ref = compiled(args)
torch.cuda.synchronize() # Ensures that segfaults are surfaced
```

This throws our device asserts
```
<frozen importlib._bootstrap_external>:843: _call_with_frames_removed: block: [45,2,0], thread: [32,0,0] Assertion `index out of bounds: 0 <= tmp4 < 24` failed.
```

I checked our generated code, and I think our device asserts might not be correct. In particular, I suspect that the mask here is not corresponding to the right mask (should be ymask instead of xmask I think). Although I'm pretty unconfident about this.

image

### Error logs

_No response_

### Minified repro

_No response_

### Versions

N/A

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #100530

When indices for indirect read are computed rather than read from another tensor, they should be masked according to the index used in computation. Currently though we don't associate masks with index variables, so the computed indices don't have associated masks also. This PR associates masks with index variables when they are created. 

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @lezcano, the example in the test doesn't generate `device_assert` on main, even though I thought it should?
On this PR, both the device assert and masked load are generated, and hopefully device assert should be removed later once your value analysis PR lands.

### CODE REVIEW

**Review by Chillee — APPROVED** [review]

**Review by lezcano — APPROVED** [review]

**Comment by lezcano** [comment]
This example should indeed have a `device_assert` in master. The issue why it's not having it (but it does after this PR) is because before this PR, the `mask` that was being passed to `load` was `None`. I (wrongly) assumed that, if the `mask` was `None` it was because we somehow knew that the indices were correct. Now, the `mask` really was `None` because of this bug lol

**Comment by ngimel** [comment]
Ok I think this is a problem. I didn't double check but I think we still have @fdrocha's optimization where we won't generate masks if `xnumel` is multiple of the biggest possible block in config? If that's the case, it's totally possible to read indices from an input in an unmasked way and then use them for indirect access, again, unmasked. If there's no `device_assert` generated in this case it's possible that invalid indices will slip through. @lezcano can you please check if that's happening?
